### ISSUE TITLE
Unable to open for writing: failed to write string

### ISSUE THREAD

**Issue by MrSmite** [yt-dlp/yt-dlp#3470]
### Checklist

- [X] I'm reporting a bug unrelated to a specific site
- [X] I've verified that I'm running yt-dlp version **2022.04.08** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Description

OS: Win 7 x64
yt-dlp: 2022.04.08
ffmpeg: 5.0.1-full_build-www.gyan.dev

I have a root folder with full permissions. In this folder, yt-dlp creates a subfolder based on the YouTube uploader ID (pulled from the URL) and then begins downloading "bestaudio", "bestvideo". Recently though I've been getting the following error:

```
ERROR: unable to open for writing: Failed to write string
Traceback (most recent call last):
  File "yt_dlp\downloader\http.py", line 267, in download
  File "yt_dlp\downloader\common.py", line 276, in report_destination
  File "yt_dlp\downloader\common.py", line 163, in to_screen
  File "yt_dlp\YoutubeDL.py", line 815, in to_screen
  File "yt_dlp\YoutubeDL.py", line 798, in _write_string
  File "yt_dlp\utils.py", line 2096, in write_string
  File "yt_dlp\utils.py", line 2080, in _windows_write_string
OSError: Failed to write string
```

There is no other verbose information. The strange thing is the folder contains a zero-byte mp4 which I watched yt-dlp create. After the error, it continues on the playlist and creates the next video as a zero-byte file.

### Verbose log

```shell
ERROR: unable to open for writing: Failed to write string
Traceback (most recent call last):
  File "yt_dlp\downloader\http.py", line 267, in download
  File "yt_dlp\downloader\common.py", line 276, in report_destination
  File "yt_dlp\downloader\common.py", line 163, in to_screen
  File "yt_dlp\YoutubeDL.py", line 815, in to_screen
  File "yt_dlp\YoutubeDL.py", line 798, in _write_string
  File "yt_dlp\utils.py", line 2096, in write_string
  File "yt_dlp\utils.py", line 2080, in _windows_write_string
OSError: Failed to write string
```

**Comment by pukkandan** [comment]
What is the command you ran? There is no other output before the error message?

**Comment by MrSmite** [comment]
> What is the command you ran? There is no other output before the error message?

The command I used is from a script I wrote:

``
!_downloadApp! --cache-dir "!_cacheDir!" --download-archive "!libFile!" --proxy !_proxyInf! --ffmpeg-location !_ffmpegApp! -f "!theFormat!" !playlistOpt! !cmdLine! -o "!outputDir!" "!theURL!"
```

All the variables you see are constants except outputDir which is built based on what youtube returns from querying the channel ID plus a subfolder if it's a playlist or single file.

theFormat is as follows:

```
SET theFormat=^(bestvideo[ext=mp4][height=1080]+bestaudio[ext=m4a]^) / ^(bestvideo[ext=webm][height=1080]+bestaudio[ext=webm]^) / ^(bestvideo[ext=mp4][height=720]+bestaudio[ext=m4a]^) / ^(bestvideo[ext=webm][height=720]+bestaudio[ext=webm]^) / ^(bestvideo[ext=mp4][height=480]+bestaudio[ext=m4a]^) / ^(bestvideo[ext=webm][height=480]+bestaudio[ext=webm]^) / best[ext=mp4] / best[ext=webm] / best
```


and the only other information before the error was this:

Downloading webpage
Downloading android player API JSON
Downloading 1 format(s): 137+140
ERROR: unable to open for writing: Failed to write string


I've been using my script for a while, since you last helped me with the channel id issue. I think the problem is when the channel has shorts mixed in with regular videos.

**Comment by dirkf** [comment]
Might there be a character in this channel name that is not acceptable in a Windows filename?

**Comment by pukkandan** [comment]
> and the only other information before the error was this:
Downloading webpage
Downloading android player API JSON
Downloading 1 format(s): 137+140
ERROR: unable to open for writing: Failed to write string

Provide the ENTIRE VERBOSE log like the issue template asked you to!!

**Comment by MrSmite** [comment]
> Provide the ENTIRE VERBOSE LOG like the issue template asked you to!!

You may have noticed I mentioned there is no additional verbose output. When I use the --verbose option I get no additional information, no log file, nothing.

> Might there be a character in this channel name that is not acceptable in a Windows filename?

Yes, I think it is happening with channels that have emojis in the name. But what I find strange is that yt-dlp creates the mp4 and m4a files in the correct path but it then fails to write to them.

**Comment by MrSmite** [comment]
Ok, so this is interesting... if I right-click on yt-dlp.exe and set it to "run as administrator", I no longer get the "failed to write string" error.

Question is why do I now get this error when I have been using yt-dlp for a while without problems?

**Comment by Lesmiscore** [comment]
We have nothing to discuss without entire verbose log. You even cropped off video IDs.

Paste the log starting from like this:
```
$ yt-dlp -v
[debug] Command-line config: ['-v']
[debug] Encodings: locale UTF-8, fs utf-8, out utf-8, err utf-8, pref UTF-8
[debug] yt-dlp version 2022.03.08.2 [e491d06d3]
[debug] Python version 3.9.12 (CPython 64bit) - Linux-5.15.5-051505-generic-x86_64-with-glibc2.34
[debug] exe versions: ffmpeg n4.4.1-3-ge16ff82624-20211127 (setts), ffprobe n4.4.1-3-ge16ff82624-20211127, rtmpdump 2.4
[debug] Optional libraries: Crypto, mutagen, sqlite, websockets
[debug] Proxy map: {}
Usage: yt-dlp [OPTIONS] URL [URL...]

yt-dlp: error: You must provide at least one URL.
Type yt-dlp --help to see a list of all options.
```

**Comment by MrSmite** [comment]
> We have nothing to discuss without entire verbose log. You even cropped off video IDs.

As I said before, I used the --verbose option and what I posted was what was in my console. I didn't get any debug lines like you show. I will try running it on a different machine to see if I get different results.

I'm curious though, if the latest yt-dlp was compiled with python 3.9.12, that is not compatible with Win 7, could that be the issue? Which version of yt-dlp uses python 3.8.xn the last Win 7 supported python?

**Comment by pukkandan** [comment]
> The command I used is from a script I wrote:
> 
> ```
> !_downloadApp! --cache-dir "!_cacheDir!" --download-archive "!libFile!" --proxy !_proxyInf! --ffmpeg-location !_ffmpegApp! -f "!theFormat!" !playlistOpt! !cmdLine! -o "!outputDir!" "!theURL!"
> ```

This command does not have `--verbose` added, and 

> Downloading webpage
> Downloading android player API JSON
> Downloading 1 format(s): 137+140
> ERROR: unable to open for writing: Failed to write string

this output is indicative of not being in verbose mode (plus you have edited these lines - don't do that!)

> ERROR: unable to open for writing: Failed to write string
Traceback (most recent call last):
  File "yt_dlp\downloader\http.py", line 267, in download
  File "yt_dlp\downloader\common.py", line 276, in report_destination
  File "yt_dlp\downloader\common.py", line 163, in to_screen
  File "yt_dlp\YoutubeDL.py", line 815, in to_screen
  File "yt_dlp\YoutubeDL.py", line 798, in _write_string
  File "yt_dlp\utils.py", line 2096, in write_string
  File "yt_dlp\utils.py", line 2080, in _windows_write_string
OSError: Failed to write string

while in this log, you haven't given the lines with `Downloading webpage` etc

Which is why I believe there must be something missing from the logs you provided. Are you saying you don't get those lines if you use `-v`, but do get them when you don't?

What do you get if you run just `yt-dlp -Uv`?


> I'm curious though, if the latest yt-dlp was compiled with python 3.9.12, that is not compatible with Win 7, could that be the issue? Which version of yt-dlp uses python 3.8.x, the last Win 7 supported python?

The official builds are made with 3.8, specificaly for compat with win7. Are you using a custom build?

**Comment by MrSmite** [comment]
Following up for anyone who finds this thread in the future. The problem wasn't the CHCP 65001 as I initially thought but rather simply the default console font wasn't a TrueType font.

I couldn't understand why unicode codepage wouldn't handle the emoji characters. I wasn't expecting them to print properly but I thought I'd at least get little boxes in place of the emoji. Instead I kept getting "couldn't write to the device" errors.

After checking folder permissions, I figured the "device" referred to must be the console's STD::OUT (or the screen). It wasn't until I ran the script on another machine that I noticed my console looked different.

**The solution:**

1. Use CHCP 65001 to enable unicode (for foreign languages, emoji, etc)
2. Right-click the console's title bar, open the _Properties_ and **change the font** to _Lucida Console_ (a TrueType font)

After making this change, I can ECHO variables holding text containing emojis to the screen (with boxes instead of emojis) and create files/folders using the same variables (windows actually creates some emojis, others are still boxes).

It still doesn't explain why the internal parts of the script (the non-ECHO stuff) didn't work. Why **_SET myvar < file.txt_** would result in _myvar_ being empty if _file.txt_ had an emoji in it. Changing the font shouldn't change how the pipe works (IMO). But it works so who cares, I guess.
